README.rst


.. image:: https://travis-ci.org/sonofeft/GenericF2PY.svg?branch=master
        
.. image:: https://img.shields.io/static/v1?label=python&message=2.7, 3.5, 3.6, 3.7&color=blue

.. image:: https://img.shields.io/pypi/l/GenericF2PY.svg


**Genericf2Py Demostrates How To Handle Fortran Binaries In Python Modules.**

See the Code at: `<https://github.com/sonofeft/GenericF2PY>`_

See the Docs at: `<http://genericf2py.readthedocs.org/en/latest/>`_

See PyPI page at:`<https://pypi.python.org/pypi/genericf2py>`_


Overview
--------

GenericF2PY focuses on the use of  `f2py <https://numpy.org/devdocs/f2py/python-usage.html>`_
for integrating FORTRAN modules into python `PyPI <https://pypi.org/>`_ projects.


GenericF2PY has the following goals:

- Support a variety of platforms Windows/Linux/MacOS/etc.
- Support a variety of python distributions (2 & 3).
- Support both 32 & 64 bit installations.

Based on my experience developing GenericF2PY, I believe the best way to achieve the above goals is:

* Distribute only **source code** (Python & FORTRAN).
* Provide guidance for setting up `f2py <https://numpy.org/devdocs/f2py/python-usage.html>`_ on the user's platform.
* Recommend users install with **"pip install <packagename>"**.
* Provide guidance for correcting any error messages.

Binary Distributions
--------------------

When binary distributions install properly, they are the most convenient approach for users... **HOWEVER**...

When binary distributions fail, they can be confounding for many users and greatly impede
or even preclude the use of your project for those users.

The sheer difficulty of maintaining binaries for all possible systems can become overwhelming.

Binary distributions can fail for a variety of reasons:

- Compiled against wrong version of `numpy <https://numpy.org/>`_
- Missing or wrong version of system libraries
- Environment variable problems (PATH, LIBRARY_PATH, CC, GCC, etc.)
- Security Restrictions
- Unknown Unknowns


It is for these reasons, that I recommend distributing **source code only**.

I'm weighing the user's difficulty of installing `f2py <https://numpy.org/devdocs/f2py/python-usage.html>`_ 
against the user's difficulty of debugging your potentially failed binary distribution and deciding that 
installing `f2py <https://numpy.org/devdocs/f2py/python-usage.html>`_ is the smaller burden.
Also, the user can Google `f2py <https://numpy.org/devdocs/f2py/python-usage.html>`_ and find answers.
You will be the user's only source of information on your binaries for all possible systems.
    
If you want to make binary distributions available, it might be best to either:

- create binary wheels in a Github **pip-wheels** directory
- provide an option to install from source code.

Users can be directed to that Github **pip-wheels** directory or given
source-install instructions as required.





HISTORY.rst
.. 2019-07-28 sonofeft 76f3109a8fc0fccdaf75777db0a44aaa67ed9ea1
   Maintain spacing of "History" and "GitHub Log" titles

History
=======

GitHub Log
----------

* Jul 30, 2020
    - (by: sonofeft) 
        - add fix for python 3.8 new os.add_dll_directory command

* Jul 28, 2019
    - (by: sonofeft) 
        - doc updates
        - Abandoned direct \*.whl download
        - upload eggs to github
        - Update UPLOAD_TO_PyPI.BAT
        - doc updates and trying direct pip github install
        
* Jul 24, 2019
    - (by: sonofeft) 
        - working multiple OS installs
        
* Jul 23, 2019
    - (by: sonofeft) 
        - start ubuntu setup
        
* Jul 21, 2019
    - (by: sonofeft) 
        - Update .travis.yml
        
* Jul 20, 2019
    - (by: sonofeft) 
        - Update .travis.yml
        - travis ci check

* Jul 12, 2019
    - (by: sonofeft)
        - First Created GenericF2PY with PyHatch


setup_v2.py
"""A setuptools based setup module for GenericF2PY

See:
http://genericf2py.readthedocs.org/en/latest/
https://github.com/sonofeft/GenericF2PY

If installing from source, then
the best way to install GenericF2PY is to use pip after navigating to the source directory::

    cd <path to where setup.py is located>
    pip install -e .

This will execute the setup.py file and insure that its pip-specific commands are run.

"""


# Use numpy to build the f2py fortran extension
# --------------------------------    
import setuptools
from numpy.distutils.core import Extension, setup


# Always prefer setuptools over distutils
try:
    from setuptools import  find_packages
except ImportError:
    from distutils.core import  find_packages

# To use a consistent encoding
from codecs import open
from os import path
import sys
                       
here = path.abspath(path.dirname(__file__))

# Get the long description from the relevant file
with open(path.join(here, 'README.rst'), encoding='utf-8') as f:
    long_description = f.read()

# Place install_requires into the text file "requirements.txt"
with open(path.join(here, 'requirements.txt'), encoding='utf-8') as f2:
    requires = f2.read().strip().splitlines()

target_file = path.join( here, 'genericf2py','_version.py')
exec( open( target_file ).read() )  # creates local __version__ variable


def _do_setup( try_to_compile=False ):
    """
    wrap the setup call so that the FORTRAN compile can be made optional.
    """
    
    if try_to_compile:
        ext_fprog = Extension(name = 'genericf2py.fprog',
                    sources = ['genericf2py/src/fprog.f'],
                    extra_link_args=["-static", "-shared"])

        # will include optional named input parameter
        optD = {'ext_modules' : [ext_fprog]}
    else:
        optD = {}

    setup(
        name='genericf2py',
        version = __version__,  # METADATA_RESET:    version = '<<version>>',
        #ext_modules = inp_ext_modules, # may be in optD

        description = 'GenericF2PY demostrates how to handle FORTRAN binaries in python modules.',
        long_description = long_description,
        long_description_content_type='text/x-rst',

        # The project's main homepage.
        url='http://genericf2py.readthedocs.org/en/latest/',
        download_url='https://github.com/sonofeft/GenericF2PY',

        # Author details
        author = 'Charlie Taylor',
        author_email = 'cet@appliedpython.com',

        # license
        license = 'GPL-3',

        classifiers = [
            # Common status values are: "3 - Alpha", "4 - Beta", "5 - Production/Stable"
            'Development Status :: 4 - Beta',

            "Operating System :: OS Independent",
            'Intended Audience :: Developers',
            "Intended Audience :: End Users/Desktop",
            'Topic :: Software Development :: Build Tools',

            # This license should match "license" above
            'License :: OSI Approved :: GNU General Public License v3 (GPLv3)',

            # Specify the Python versions you support here. In particular, ensure
            # that you indicate whether you support Python 2, Python 3 or both.
            'Programming Language :: Python :: 2',
            'Programming Language :: Python :: 2.7',
            'Programming Language :: Python :: 3',
            'Programming Language :: Python :: 3.5',
            'Programming Language :: Python :: 3.6',
            'Programming Language :: Python :: 3.7',
        ],

        platforms = 'any',

        # What does your project relate to?
        keywords = 'genericf2py setuptools development',

        packages = find_packages(exclude=['.tox', '.hg', 'docs']),
        package_data = {'genericf2py':['examples/*.*', 'src/*.*', '*.pyd', '*.so']},
        include_package_data=True,

        # List run-time dependencies here.  These will be installed by pip when
        # your project is installed.
        install_requires = requires,  # read from requirements.txt

        tests_require = ['nose','coverage'], # ...OR... ['pytest','pytest-cov']
        test_suite='genericf2py.tests', # allows "setup.py test" to work

        # List additional groups of dependencies here (e.g. development
        # dependencies). You can install these using the following syntax,
        # for example:
        # $ pip install -e .[dev,test]
        extras_require = {
            'dev': ['check-manifest'],
            'test': ['coverage'],
        },

        zip_safe= False,

        # To provide executable scripts, use entry points in preference to the
        # "scripts" keyword. Entry points provide cross-platform support and allow
        # pip to create the appropriate form of executable for the target platform.

        #entry_points = {
        #    'console_scripts': [
        #        'genericf2py=genericf2py.main:main',
        #    ],
        #},
    **optD) # may include optional named parameter

if 'build' in sys.argv:
    print('Running setup... COMPILING FORTRAN')
    _do_setup( try_to_compile=True )
else:
    print('Running setup WITHOUT compiling FORTRAN')
    _do_setup( try_to_compile=False )
    
    try:
        # see if compiled module is available.
        from genericf2py import fprog
    except:
        print('WARNING... COMPILED FORTRAN FAILED... ATTEMPTING COMPILE')
        print('Re-Running setup... COMPILING FORTRAN')
        _do_setup( try_to_compile=True )
    
    
    

